### ISSUE TITLE
Remove Instrumentation

### ISSUE THREAD

**Issue by nathanila** [jacoco/jacoco#782]
### Steps to reproduce

JaCoCo version: 0.8.2
Operating system: Windows 10
Tool integration: -javaagent on the command line to generate the Exec file and Jacoco Java code to generate the CSV and HTML reports

### Actual behaviour

Running with Jacoco adds significant execution (CPU) overhead.  I assume that the Jacoco Java agent instruments the bytecode and then leaves the instrumentation in place the entire time.

### Expected behaviour

I propose that the instrumentation bytecode is removed as the instruction, branch, line or method is completely covered.  For example, let's say an instruction executes for the first time, then Jacoco removes the instrumentation for that instruction.  By the time JIT optimizes the method, Jacoco has removed most if not all of the instrumentation.  This means frequently executed instructions will not have any Jacoco instrumentation and hence run at full speed.

I realize that removing the instrumentation could be expensive.  The instrumentation should be removed in a background thread on a delay.  This way the instrumentation for a class is removed with 1 or a few class redefinitions.

Due to the limitations of class redefinition, the only thing that can be adjusted is the method bodies.  This should be adequate since unused fields do not cause a CPU overhead.

I do not know how Jacoco works internally.  But let's say this is the code after instrumentation.  The realMethod() is the code under test.  markMethod() is Jacoco code to mark the method as covered.

```
public void realMethod()
{
   markMethod();
   ... real code ...
}

private static void markMethod()
{
   ... marks the method as covered ...
}
```

After markMethod() runs the first time, then Jacoco can redefine the method to be empty.  When JIT optimizes realMethod(), the entire call to markMethod() is optimized away.  Hence, the overhead for markMethod() drops to 0.

Jacoco could remove the call to markMethod() from realMethod().  This would remove the instrumentation overhead even if JIT never optimizes the method.

I am new to Jacoco.  If Jacoco can record the method callers, then removing markMethod() is not an option since Jacoco needs to track what called readMethod().  Allow for the user to turn off any conflicting features so that instrumentation can be removed.  Inform the user why instrumenation can not be removed (i.e. some option prevents it).

**Comment by marchof** [comment]
@nathanila Can you please elaborate a bit about your use case?

- How long does your overall test run?
- How die you meassure the "significant execution (CPU) overhead"? What is the overhead %?

**Comment by nathanila** [comment]
Normally my test uses 400% CPU (i.e. 4 hyper-threads on Intel chip).  With Jacoco agent enabled, my test uses 1700% (i.e. 17 hyper-threads).  This shows more than a 4x increase in CPU usage due to Jacoco.

**Comment by marchof** [comment]
Thanks for the details! From our own measurements with standard unit tests (e.g.  for Apache commong-collections) the overhead is less than 10%. We learned that for CPU intensive tests (e.g. commons-math) the overhead can be bigger,

- So I assume your test scenario is highly multithreaded, right?
- What is the nature of your application? Is it primarly CPU intensive calculations?

**Comment by nathanila** [comment]
I am using TestNG and run 50 test cases concurrently.  I have up to 350 threads running but the typical thread count is around 100.

The test cases download gigabytes of data, parse the data and display in Java Swing tables.  Hundreds of tables are displayed.

**Comment by nathanila** [comment]
I captured a Java Flight Recorder profile.  I looked at the hottest method.  This method is called multiple times for every character in the JSON strings in the gigabytes of data (think String.charAt()).  I then used javap to look at the bytecode before and after instrumentation.

I see that Jacoco creates a byte array and puts it in a static field.  Then as the method executes, the added bytecode sets elements in the array to 1.  In a multi-threaded scenario, this is going to really hurt cache performance.  The cache line holding the elements of the byte array is going to travel to the various caches in the cores.  Think of it as a spin lock placed around every write to the byte array.

The HotSpot JVM ran into a similar problem.  When a thread updates a field, then a flag (i.e. card mark) has to be set so that GC can know where writes were made in the heap.  At first, every update would set the card mark to 1.  This caused performance issues like what we are seeing with Jacoco.  The JVM team then switched to checking if the card mark is set to 0 and if so, then set the card mark to 1.

I recommend you do a similar thing.  This will add a branch bytecode but the good thing is that JIT will see that the branch always evaluates the same way and minimize the impact in the machine code.  The processor will see that the branch always evaluates the same way and predict correctly.  Thus, the impact of the branch will be very small (i.e. 6 cycles - 3 to load the value and 3 to compare).  Performance will greatly improve!

Currently, the Jacoco bytecode looks something like this in Java.
`
jacocoData[3] = 1;
`
Each time the statement is executed, the thread will pull the cache line into the core in the shared state.  The thread will invalidate the cache line in the rest of the cores and change the cache line to the exclusive state.  The thread will write a 1 to the cache line... even if jacocoData already has a 1.  All of the threads will do these operations.  The threads will contend to put the cache line in the exclusive state.  As the thread waits, it will look like it is spinning in a spin lock (i.e. high CPU usage).

I propose you change the Jacoco bytecode to look something like this in Java.

`
if (jacocoData[3] == 0)
   jacocoData[3] = 1;
`

Each time the `if` statement is executed, the thread will pull the cache line into the core in the shared state.  The thread will then execute the branch.  The first time this code executes, the `jacocoData[3] = 1;` will execute.  This has the same penalties as mentioned in the previous case.  The subsequent times the `if` statement executes, the cache line is never invalidated.  Thus the cache line can be shared in all of the cores and performance is great!

I doubt I am the first to propose this.  Has this been tried before?  If so, what caused the idea to be rejected?

**Comment by nathanila** [comment]
Have you considered putting the call to jacocoInit() as the first thing in the static initializer and remove the calls to jacocoInit() in all other methods?  This will help reduce the size of the JIT optimized machine code.

**Comment by nathanila** [comment]
I looked through Jacoco code.  I think the code change for adding the `if (jacocoData[3] == 0)` will happen in `ProbeInserter.insertProbe()`.

**Comment by nathanila** [comment]
If you create an agent jar with this code change, I am willing to check it out and see if it performs better.

**Comment by Godin** [comment]
@nathanila first of all thank you for detailed explanations and ideas.

----

> For my use case, removing instrumentation from many base classes will reduce the instrumentation overhead. I would guess that Jacoco should have a configuration option to enable removal of instrumentation.

I'm wondering if you was really looking for removal, because already possible to exclude some classes from instrumentation - see agent option "excludes" https://www.jacoco.org/jacoco/trunk/doc/agent.html

----

> Have you considered putting the call to jacocoInit() as the first thing in the static initializer and remove the calls to jacocoInit() in all other methods? This will help reduce the size of the JIT optimized machine code.

Unfortunately other methods can be executed before static initializer - see [BadCycleClassTest](https://github.com/jacoco/jacoco/blob/v0.8.2/org.jacoco.core.test.validation.java5/src/org/jacoco/core/test/validation/java5/BadCycleClassTest.java)

----

> I doubt I am the first to propose this. Has this been tried before?

AFAIR you're not the first about static initializers, but the first in both - about problem statement in case of heavy multithreading and in proposal to add branch, so AFAIK we have not tried/considered this before.

About current code of probe see https://www.jacoco.org/jacoco/trunk/doc/flow.html - it is indeed `$jacocoData[id] = true` . And so requires `7` bytes at max.

If math below is correct (I'm quite tired and sleepy), then addition of branch `if ($jacocoData[id] == 0)` will more than double maximal size of probe - by 9 bytes in total: 2 bytes for `aload`, 3 bytes to push `id`, 1 byte for `baload`, 3 bytes for `ifne`. And so will negatively impact inlining - `MaxInlineSize` and `FreqInlineSize`. Also will increase chances to reach 64KB method size limit.

Can think about placement of code of probe into separate method - 6 bytes in total: 3 bytes to push `id`, 3 bytes for `invokestatic`. However this also will impact inlining - `MaxInlineLevel`.

Also wondering about `InlineSmallCode` as all this affects native code.

And so wondering about impact on tests that are far less highly multithreaded. BTW how reduction of threads impacts your tests? Wondering if possible to write JMH benchmark or find some other way that  can demonstrate what you observe.

Initial idea about removal of instrumentation by class redefinition IMO has even more aspects and pitfalls for consideration.

**Comment by Godin** [comment]
Additional thought about `MaxInlineLevel`: we anyway add `$jacocoInit`.

---

And here is attempt to write benchmark:

```java
import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.State;

@State(Scope.Thread)
public class Bench {

	int x;

	static boolean[] probes;

	private static boolean[] init() {
		if (probes == null)
			probes = new boolean[100];
		return probes;
	}

	@Benchmark
	public int original() {
		return x;
	}

	@Benchmark
	public int simulated_current() {
		boolean[] probes = init();
		probes[0] = true;
		return x;
	}

	@Benchmark
	public int simulated_proposed_branch() {
		boolean[] probes = init();
		if (probes[0] == false)
			probes[0] = true;
		return x;
	}

	@Benchmark
	public int simulated_proposed_branch_in_method() {
		init();
		hit(0);
		return x;
	}

	private static void hit(int id) {
		if (probes[id] == false)
			probes[id] = true;
	}

}
```

```
1 thread
Benchmark                            Mode  Cnt  Score   Error  Units
original                             avgt    5  2.327 ± 0.011  ns/op
simulated_current                    avgt    5  2.842 ± 0.145  ns/op
somulated_proposed_branch            avgt    5  2.800 ± 0.131  ns/op
simulated_proposed_branch_in_method  avgt    5  2.782 ± 0.005  ns/op

2 threads
Benchmark                            Mode  Cnt   Score   Error  Units
original                             avgt    5   2.383 ± 0.206  ns/op
simulated_current                    avgt    5  15.318 ± 0.645  ns/op
simulated_proposed_branch            avgt    5   2.796 ± 0.078  ns/op
simulated_proposed_branch_in_method  avgt    5   2.813 ± 0.094  ns/op

4 threads
Benchmark                            Mode  Cnt   Score   Error  Units
original                             avgt    5   2.499 ± 0.297  ns/op
simulated_current                    avgt    5  39.492 ± 4.186  ns/op
simulated_proposed_branch            avgt    5   3.021 ± 0.172  ns/op
simulated_proposed_branch_in_method  avgt    5   3.012 ± 0.233  ns/op

8 threads
Benchmark                            Mode  Cnt   Score   Error  Units
original                             avgt    5   4.802 ± 0.077  ns/op
simulated_current                    avgt    5  75.675 ± 0.271  ns/op
simulated_proposed_branch            avgt    5   5.740 ± 0.067  ns/op
simulated_proposed_branch_in_method  avgt    5   5.726 ± 0.102  ns/op

16 threads
Benchmark                            Mode  Cnt    Score   Error  Units
original                             avgt    5    9.595 ± 0.335  ns/op
simulated_current                    avgt    5  130.283 ± 3.181  ns/op
simulated_proposed_branch            avgt    5   12.760 ± 2.878  ns/op
simulated_proposed_branch_in_method  avgt    5   12.247 ± 2.643  ns/op

32 threads
Benchmark                            Mode  Cnt    Score    Error  Units
original                             avgt    5   19.995 ±  1.331  ns/op
simulated_current                    avgt    5  328.762 ± 23.489  ns/op
simulated_proposed_branch            avgt    5   23.653 ±  1.119  ns/op
simulated_proposed_branch_in_method  avgt    5   24.368 ±  1.969  ns/op

64 threads
Benchmark                            Mode  Cnt    Score    Error  Units
original                             avgt    5   41.846 ±  3.774  ns/op
simulated_current                    avgt    5  706.047 ± 38.067  ns/op
simulated_proposed_branch            avgt    5   53.071 ±  5.956  ns/op
simulated_proposed_branch_in_method  avgt    5   46.426 ±  0.950  ns/op
```

**Comment by nathanila** [comment]
My first Jacoco optimization was to exclude as much as possible.  This helped since I removed all of the test methods.  Thanks for the pointer though.

Thanks for the pointer to BadCycleClassTest.  That was interesting.

As for inlining, you could add documentation that it might improve performance to adjust the JVM's inlining thresholds.  Thus, methods that were inlined before Jacoco can be inlined after Jacoco.

I didn't think about the 64 KB method size limit.  There is no workaround for that since bytecode does not allow us to override this constraint.  Perhaps, Jacoco needs a configuration option to select between compact (original code) versus performance (branch code).

The current best practices on method design is to keep the methods focused and do only 1 thing and to keep the method small.  So, 64 KB methods should be rare.  I just heard a story of a method with 40,000 lines of code.  A person is assigned to refactor the method into many small methods.  In 1 year's time, the person managed to get the method down to 30,000 lines of code.  I guess the 64 KB method size limit is a real concern.

For tests that do not warm code, the performance is going to degrade by 6 cycles per probe versus the current solution.  3 cycles to load the value from jacocoData and 3 cycles to compare the value to 0.  (Cycles for Intel processors.)

I agree that the removal of instrumentation by class redefinition is going to be tricky.  The code itself cannot do this since this would create a lot of code.  Thus, a background thread will have to poll the jacocoData in all of the classes, detect newly covered probes and then redefine the class without the probes.  For code that executes only 1 time, this wastes CPU time.

I like the benchmark code.  It looks like it captures our ideas very well.  From the results, it appears that you ran this on a machine with 4 physical cores.  I deduce this from the fact that original benchmark doubles its average at 8 threads.  If the machine has hyper-threading, it won't help at 8 threads since the cores are executing a compute intensive workload and hyper-threading cannot help here.

The results show that simulated_proposed_branch_in_method is definitely better than the current solution for all numbers of threads.  For 2+ threads, I already gave the answer.  For 1 thread, this is a bit tougher since inlining should make simulated_proposed_branch and simulated_proposed_branch_in_method appear the same in machine code.  I would be curious to look at the assembly code for these two benchmarks, but it probably is not important.

simulated_proposed_branch_in_method adds the following overhead compared to original.  The error is much higher at 16 threads.  I am wondering if something else was running on the machine.  I am also wondering why the error is so much lower for simulated_proposed_branch_in_method with 64 threads.

1 thread: 19.6%
2 threads: 18.0%
4 threads: 20.5%
8 threads: 19.2%
16 threads: 27.6%
32 threads: 21.8%
64 threads: 10.9%

I am wondering if you can run more counts (test iterations) on simulated_proposed_branch and simulated_proposed_branch_in_method for 1, 2 and 4 thread counts.  I am curious if a clear winner emerges.  If no clear winner emerges, then I suggest implementing simulated_proposed_branch_in_method.  This adds fewer bytecodes per probe.

Microbenchmarking shows that adding a branch is going to help.  If you create an agent jar on version 0.8.2 with branches added, then I can see how it performs on my code.

**Comment by nathanila** [comment]
> For tests that do not warm code, the performance is going to degrade by 6 cycles per probe versus the current solution. 3 cycles to load the value from jacocoData and 3 cycles to compare the value to 0. (Cycles for Intel processors.)

If the code is JIT optimized, then the performance impact is going to be 6 cycles per probe.  However, I made the wrong assumption that the code would be JIT optimized.  Instead, the interpreter is going to be involved.  I do not know how much overhead the interpreter adds for an array load and a branch.

**Comment by marchof** [comment]
I tried to hack a quick prototype. Unfortunatelly it is not that easy than is looks like: The new condition requires us to add a frame. We currently don't have proper frame information at hand.

**Comment by Godin** [comment]
@nathanila 

> From the results, it appears that you ran this on a machine with 4 physical cores.

Yes - this result is from my workstation at work, where use Mac 2.8 GHz Intel Core i5 with 4 physical cores.

> I am wondering if something else was running on the machine.

This is just a sketch, quick attempt, definitely not a clean lab - maybe IDE, Chrome, Zoom or something else also were consuming CPU.

> I am wondering if you can run more counts (test iterations)

Definitely needs more careful run.

> I would be curious to look at the assembly code for these two benchmarks

AFAIR `perfasm` in JMH (at least in past) wasn't available on macOS.

I plan to do more careful and clean run at home on Linux.

----

@marchof 

> The new condition requires us to add a frame. We currently don't have proper frame information at hand.

This is for case when branch is directly added into instrumented code? IMO if probe is method invocation, then we know exact frame for branch in this method.

**Comment by marchof** [comment]
@Godin Right! Adding the extra method will only work for newer Java versions in case of interfaces. But for the tests we can ignore support for older Java versions.

@nathanila Are you testing on >= Java 8?

**Comment by nathanila** [comment]
@marchof I am testing on Java 10

**Comment by Godin** [comment]
I guess that Java 10 will throw exception and use original class definition in case if transformation adds method and bytecode version < 8.

**Comment by nathanila** [comment]
@Godin Thanks for the zip file.  I will try it out.  It will take a few hours.

**Comment by Godin** [comment]
> As for inlining, you could add documentation that it might improve performance to adjust the JVM's inlining thresholds.

Idea is interesting. Unfortunately many people don't bother reading documentation and expect everything to magically work without any configuration and overheads 😢 Also many people have troubles with [just addition of `-javaagent:`](https://groups.google.com/d/msg/jacoco/_n7hMYaRasE/VA4jdUYTBQAJ), [change of `-cp`](https://groups.google.com/d/msg/jacoco/QdMLz3Wo7pg/6zfC2D3BBAAJ) and other configuration options, imagine addition of several `-XX:`  🤯 😆

I personally more interested / wondering if we can find a way to quantify impact by default, and thus to quantify effects of tuning, which will allow to experiment with tuning effects at least for ourselves.

----

> I guess the 64 KB method size limit is a real concern.

Unfortunately big methods are real - https://groups.google.com/d/msg/jacoco/RvYmvpeHMl8/kl90La-SCwAJ , https://groups.google.com/d/msg/jacoco/vuudASfN1yg/T-LpSRUdAgAJ , #177 and even `javac` generates them - see https://github.com/jacoco/jacoco/pull/462#issuecomment-266293585.

**Comment by nathanila** [comment]
@Godin Thanks for the pointers.  This conversation has been enlightening.

**Comment by nathanila** [comment]
Here are the awesome results of running with and without Jacoco.  The CPU is the average CPU usage for the entire test time.  100% means 1 core fully utilized.  The GC is the total amount of time spent in GC.

No Jacoco:  CPU: 430% - GC: 1:14.524
Jacoco with Godin's change:  CPU: 498% - GC: 1:10.289

This is way better!  Once this is released, Jacoco users will no longer be severely impacted in multi-threaded tests.

One might be tempted to say that Jacoco causes an increase of 16% (= 498 / 430) in CPU and a decrease of 5.6% in GC.  However, my tests have a lot of variance in CPU and GC time.  Just today without Jacoco, I saw a CPU usage of 503.6%.  I have seen GC in the 2 minute range.  So, we cannot be sure how much impact Jacoco is causing.

**Comment by nathanila** [comment]
The code uses a boolean array to store the coverage.  Have you considered using static byte fields to store the coverage?  (Do not use boolean fields since these are encoded as int fields and would consume 4x more space.)  By using a boolean array, the code incurs a memory indirection to read/write the coverage.  By using byte fields, there is no memory indirection.  Here is what the code might look like.

```
...
   probe3();
...

private static void probe3() {
   if (s_covered3 == 0)
      s_covered3 = 1;
}
```

We may not see any impact in the micro-benchmark since the L1 cache will have the array.  On the other hand, JIT may be able to hardcode the address of the field in the load/store instructions and achieve better performance than accessing the array.  Can you try this in the micro-benchmark?

The above code will require 3 bytes in the method per probe.  The downside is that every probe will have its own static method.  I am not sure how big of a problem this is.  There is a limit of 64 K methods in 1 class which means there would be a limit of 64 K probes in 1 class.  If the method limit is some concern, the probe limit seems to be a significant concern.

I thought about putting the probe fields and methods in separate classes (up to the bytecode limits).  Jacoco could then define any number of these classes.  The problem is that calling a static method requires using a slot in the constant pool.  The constant pool has a limit of 64 K per class.  So, whether the static method is in the same class or a different class, it isn't going to make much difference.  The constant pool will be the limiting factor.

I suppose we could take a hybrid approach.  Use byte fields until we are _close_ to hitting a bytecode limit.  Then use the boolean array for the rest of the probes.

**Comment by Godin** [comment]
> The code uses a boolean array to store the coverage. Have you considered using static byte fields to store the coverage?

Yes we considered: array-based probes up to 32767 do not require constant pool entry, whereas field/method-based would require - https://twitter.com/_godin_/status/889089219309821952

> I suppose we could take a hybrid approach. Use byte fields until we are close to hitting a bytecode limit.

Can't - there should be space for other bytecode manipulation tools that run after us. Also hybrid is/will be harder to maintain.

**Comment by nathanila** [comment]
Thank you for the insights.  It seems calling a static method and conditionally setting the boolean array element to true is the best solution (i.e. simulated_proposed_branch_in_method).

Thank you for the effort and making Jacoco usable for my situation!  I'll keep using the special version you gave me until 0.8.3 is released.

**Comment by marchof** [comment]
To simplify things: I wonder what the overhead would be if we combine `$jacocoInit` and `$jacocoHit` into a single method. For small methods this could improve performance because a) we have method call less and b) we can directly use the probe array loaded from the static field.

**Comment by Godin** [comment]
Thinking more about `$jacocoInit`: I guess that we already hurt inlining, because it calls our runtime. Thus wondering maybe better to sacrifize precision in case of bad cycle - move `$jacocoInit` into static initializer and simply add null-check into `$jacocoHit` ?

**Comment by nathanila** [comment]
@Godin I recommend running a JMH test to see how performance is impacted.  The null check will be very predicatable and will probably only add 3 cycles.

If a method has a call to `$jacocoInit`, this will only hurt inlining due to the larger method (3 bytes?).  This method can still be inlined and when inlined it can still call `$jacocoInit`.  `$jacocoInit` does not have to be inlined in order for the method to be inlined.

What about using an initialization-on-demand holder idiom?

```
private static class jacocoHolder {
   private static final boolean jacocoData[] = new boolean[5];
}
```

This completely avoids bad cycle since `jacocoHolder` does not derive from anything that could cause bad cycle.  It also lets us declare `jacocoData` as final which may help JIT.  I have not looked at the bytecode for accessing `jacocoHolder.jacocoData`, so I am not sure if this is going increase the bytecode versus no holder.  I have also not looked at the assembly code (JIT output) for this either.

**Comment by marchof** [comment]
@nathanila We tried this "companion class" aproach before, see #426 .

**Comment by Godin** [comment]
@nathanila 

> declare jacocoData as final which may help JIT

AFAIK nowadays JVM won't trust our `static final` field - see https://shipilev.net/jvm-anatomy-park/17-trust-nonstatic-final-fields/

> I recommend running a JMH test to see how performance is impacted

I do plan to do this. And this requires some free time. Please take into account that we don't work on JaCoCo full-time.

> I am proposing 1 companion class per application class.

We were thinking about this too. And both cases require careful implementation with experiments and with consideration of various cases such as offline instrumentation without agent, which also takes time.

**Comment by Godin** [comment]
@nathanila Just to prevent misunderstandings: all your ideas are interesting - we value this, and big thank you for them! :heart: Some time is required for experiments and for careful consideration of all these ideas in addition to theoretical speculations. Respectively if you have this time and interest, then feel free to do experiments on your own and show the results / conclusions :wink:

**Comment by nathanila** [comment]
Before I added Jacoco, the **worst**-case CPU usage for my tests was 505%.  After I added Jacoco, the **worst**-case CPU usage increased to 550%.  That is an 8.9% increase.  It looks like multi-threaded performance is now on par with single-threaded performance.  That's awesome!  I hope this translates into a huge increase in Jacoco usage.

What does worst-case CPU usage mean?  I have one `@AfterSuite` test configuration that checks the average CPU usage for the test.  If the average goes above a threshold, then I have to decide if I should profile and optimize the code (because I added the slow version of Jacoco) or if I should live with the performance degradation (because I added the performant version of Jacoco) and increase the threshold.  Yesterday, I ran the test suite 1 time and reported to you the average CPU with and without Jacoco.  After merging in Jacoco code coverage, I ran the test suite several times and adjusted the CPU threshold.  All night and morning, I ran the test suite about 100 times.  (I have a program that automatically kicks off another test if none are running.)  The CPU threshold did not need adjusting any further.  Thus, I know that the performant version of Jacoco caused a 8.9% increase in worst-case CPU usage.

Why do I monitor the CPU usage for the test?  Let's say I make a code change in a private branch that spikes the CPU usage, then I go back to the code in the private branch and figure out how to make the code faster.  In other words, it helps me catch significant performance regressions.  I do the same for GC time and total bytes allocated.

**Comment by marchof** [comment]
@nathanila Out of curiousity: Are you allowed to share what kind of application you are working on?

**Comment by nathanila** [comment]
I work on a diagnosis application.  The server downloads and stores information about the state of several large systems of machines.  The user opens the Swing client and browses through the history of data to figure out what is going on and what happened.  The tests download about 10 GB of JSON data, process the data and display the data in Swing tables.  This ensures the program can handle anything new.  The unit tests have all of the cases I've hit before and ensure I can always handle those cases.

**Comment by nathanila** [comment]
> Thinking more about `$jacocoInit`: I guess that we already hurt inlining, because it calls our runtime. Thus wondering maybe better to sacrifice precision in case of bad cycle - move `$jacocoInit` into static initializer and simply add null-check into `$jacocoHit`?

Here is an expansion on that idea.  Bad cycle is very rare.  Furthermore, `$jacocoData` is rarely going to be null inside `$jacocoHit()`.  So, I suggest **not** checking for null and wrapping the access in a catch `NullPointerException`.  Here is some example code.

```
private static void jacocoHit(int index)
{
   while (true)
   {
      try
      {
         if (!jacocoData[index])
            jacocoData[index] = true;

         return;
      }
      catch (NullPointerException e)
      {
         jacocoInit();
      }
   }
}
```

The while loop and try-catch block cost nothing for the common case.  That is, the bytecode instructions for these constructs are all after the try block.  Thus, this method should be just as performant as `simulated_proposed_branch_in_method`, but we get rid of the calls to `$jacocoInit` in all of the methods except the static initializer!

If `jacocoData` is null, then the catch block executes `jacocoInit()` and jumps back to the if statement.  This rare case is going to be expensive.  However, it will happen only 1 time for each Bad Cycle class, a rare thing.

As for bytecode size of `jacocoHit()`, I think it will be the same as executing the null check.  The null check will add a load and null branch (6 bytes).  The above code will add a call to `jacocoInit()` and a jump (6 bytes) plus a label and catch block in the method metadata.  I don't think the method metadata counts against the inlining of the method.

The JVM knows that throwing and catching exceptions is expensive.  If the `NullPointerException` were thrown frequently, JIT will put a null check in the assembly code.  In Jacoco's case, `NullPointerException` will happen only 1 time.  Thus, the interpreter will have to deal with it.  I highly doubt it would impact the assembly code.

**Comment by nathanila** [comment]
Another advantage of calling `jacocoHit()` (whatever its implementation) is that profilers will correctly attribute the CPU time to `jacocoHit()` instead attributing the CPU time to the methods.  In other words, Jacoco overhead is made visible to profilers and hence developers.  The developers can then understand why a method is consuming so much CPU time.

**Comment by nathanila** [comment]
Oh, by the way, I have been using the modified Jacoco that you gave me.  It is working great!  I haven't had an issue with it all this time after 1000s of runs.

**Comment by nathanila** [comment]
It might make sense to have a global static method `jacocoHit(boolean jacocoData[], int index)`.

- This method will be optimized by JIT much faster than having individual methods in each class since the method will be invoked frequently.
- This won't affect inlining any more than individual methods.
- It shouldn't make bounds checking any worse since bounds checking optimization does not happen until JIT optimizes the method.  In either case, once the method is inlined, the bounds checking disappears.
- The advantage of having a single global method is that one can see the overhead of this method in profilers and hence the user can see most of Jacoco's overhead.

**Comment by marchof** [comment]
@nathanila Would simplifly a lot if we could provide global APIs to the instrumented code. Unfortunatelly we haven't found a simple way yet how to provide such an API at runtime. Think of app servers and class loaders.

See "Coverage Runtime Dependecy" in https://www.jacoco.org/jacoco/trunk/doc/implementation.html

**Comment by nathanila** [comment]
One of you delivered a Jar file for me to try out.  I have been using that Jar file ever since.  It is working great!  Has that code been merged and put into a release?  If so, what release is it in?

**Comment by Godin** [comment]
@nathanila no, this was not integrated

**Comment by anatoliy-balakirev** [comment]
Hi. Any news on this one? We experience the same issue with our build taking 26 mins without and 1 hour with Jacoco.
